daemon/container_operations.go

```diff
@@ -691,6 +691,12 @@ func (daemon *Daemon) connectToNetwork(cfg *config.Config, container *container.
 	}
 	nwName := n.Name()
 
+	// If idOrName is a partial ID, we just delete the matching entry from container's endpoint settings. An entry
+	// will be recreated below with the network name used as a key.
```

Comment:
My concern here is still that "deleting" and "re-creating" could mean "delete settings that were defined for one". Say, the `shortid` has Links or Aliases defined, but the "full name" doesn't, wouldn't that mean we'd delete the links and aliases?

https://github.com/moby/moby/blob/6ce5aa1cd5a46e02ffaa3fe0c794642b35601676/daemon/network/settings.go#L31-L37
https://github.com/moby/moby/blob/6ce5aa1cd5a46e02ffaa3fe0c794642b35601676/api/types/network/network.go#L34-L50


So in that case, we'd either;

- produce an error (already connected)
- have some way to merge the two (but "how" because we don't know which properties to merge (and should they?)

Overwriting any prior config _COULD_ be a use-case, but in that case we must define what the behavior is. And it would also (I think?) only be if we'd have `docker container update` allow updating and/or a `--force` (or something) option on `docker network connect`.

If a container would have multiple network-attachments on _create_, it MUST produce an error (trying to create a container attached to the same network multiple times)

Reply:
- Also, TIL about `IPAMOperational`. Looks like we have multiple approaches to separate operational from non-operational data (related to https://github.com/moby/moby/pull/46059)

Reply:
> My concern here is still that "deleting" and "re-creating" could mean "delete settings that were defined for one".

For the "deleting" bit, only the `EndpointSettings` for the current partial ID will be deleted. For the "re-creating" bit, this is what the existing code does; although it can't happen (in any case) thanks to the error returned by the `CreateEndpoint` call.

> If a container would have multiple network-attachments on create, it MUST produce an error (trying to create a container attached to the same network multiple times)

Even in that case, the `CreateEndpoint` call will fail:

```
$ docker network ls --filter=name=testnet1                                                                                            
NETWORK ID     NAME       DRIVER    SCOPE
f4746355e4df   testnet1   bridge    local

$ docker run --rm --name test --network=name=testnet1,ip=192.168.100.24 --network=name=f4746355e4df,ip=192.168.100.30 -d nginx
a4dc967e2167a0485c9bce45806139c27af69a8b19d69704ffd4b5ee0fc73db5
docker: Error response from daemon: endpoint with name test already exists in network testnet1.
```

Are you suggesting we should check that no two network-attachments reference the same network _before_ we call `connectToNetwork`? In that case, yeah I agree it'd be great, but:

* Since a swarm-scope network might not exist yet on the local node, if we do that before `findAndAttachNetwork`, we might not detect if two partial IDs of different size reference the same swarm-scope network
* To do that properly `allocateNetwork` (btw, it should be renamed into `allocateNetworks` to make it less confusing) should create swarm attachments and normalize/validate partial IDs.
* ^ I think this is out of scope for a fix we need to backport.

Reply:
> Are you suggesting we should check that no two network-attachments reference the same network before we call connectToNetwork? In that case, yeah I agree it'd be great, but:

I'm considering that;

- Yes, if we have an opportunity to catch invalid configs (same network included _twice_ in the container's network-attachments) that we should produce an error early
- But we should also guard against such a config getting here through other paths if possible

Basically `container.NetworkSettings.Networks` having the same network _twice_ but potentially using different identifiers (`name`, `ID`, "partial ID") "should not happen".

Now, as this code is new, perhaps there's existing configs that may have this situation (?), so maybe we currently _can't_ make this a hard failure, but in that case, we _could_ consider printing a warning; something like (dummy code);

```go
if idOrName != n.ID() && idOrName != nwName && contains(container.NetworkSettings.Networks, idOrName) {
	log.G(context.TODO()).WithFields("previous config": ....).Warn("container is connected to the same network twice. deleting previous config")
	delete(container.NetworkSettings.Networks, idOrName)
}
```